mindmeld/models/nn_utils/classification.py

```diff
@@ -0,0 +1,490 @@
+            )
+            for start_idx in t:
+                this_examples = train_examples[start_idx:start_idx + self.params.batch_size]
+                this_labels = train_labels[start_idx:start_idx + self.params.batch_size]
+                batch_data = self.encoder.batch_encode(
+                    examples=this_examples,
+                    padding_length=self.params.padding_length,
+                    add_terminals=self.params.add_terminals
+                )
+                batch_data.update({
+                    "labels": self._prepare_labels(this_labels, batch_data["split_lengths"])
+                })
+                batch_data = self.forward(batch_data)
+                loss = batch_data["loss"]
+                train_loss += loss.cpu().detach().numpy()
+                train_batches += 1
+                # find gradients
+                loss = loss / self.params.gradient_accumulation_steps
+                loss.backward()
+                # optimizer and scheduler step
+                batch_id = start_idx / self.params.batch_size
+                if (
+                    start_idx + self.params.batch_size >= len(train_examples) or
+                    (batch_id + 1) % self.params.gradient_accumulation_steps == 0
+                ):
+                    # update weights when it is the last batch in the epoch or
+                    # when specified step is reached or
+                    if self.params.max_grad_norm:  # clip (accumlated) gradients if required
+                        nn.utils.clip_grad_norm_(self.parameters(), self.params.max_grad_norm)
+                    optimizer.step()
+                    scheduler.step()
+                    optimizer.zero_grad()
+                progress_bar_msg = f"Epoch: {epoch} | Mean loss: " \
+                                   f"{train_loss / (start_idx / self.params.batch_size + 1):.4f}"
+                t.set_description(progress_bar_msg, refresh=True)
+            train_loss = train_loss / train_batches
+            # dev evaluation
+            predictions, targets = [], []
+            t = tqdm(range(0, len(dev_examples), self.params.batch_size), disable=not self.verbose)
+            for start_idx in t:
+                this_examples = dev_examples[start_idx:start_idx + self.params.batch_size]
+                this_labels_predicted = self.predict(this_examples)
+                this_labels_targetted = dev_labels[start_idx:start_idx + self.params.batch_size]
+                # validation
+                if len(this_labels_predicted) != len(this_labels_targetted):
+                    msg = f"Number of predictions ({len(this_labels_predicted)}) " \
+                          f"not equal to number of targets ({len(this_labels_targetted)})"
+                    logger.error(msg)
+                    raise AssertionError(msg)
+                # flatten if required
+                try:
+                    this_labels_predicted = sum(this_labels_predicted, [])
```

Comment:
I am unclear what this computation accomplishes

Reply:
Because the fit method is used by both sequence classification as well as sequence labeling tasks, this step is required to flatten sequence labeling task's results (i.e from [[], [], ..., []] --> [...]). Once flattened, the results data structure is same in case of both tasks and hence a generic way of scoring is possible.

Reply:
You can also use chaining from itertools (``from itertools import chain``) to achieve the same, that's supposedly faster for longer lists. I'm considering making the same switch in the AL PR. :) 

```suggestion
                    this_labels_predicted = chain.from_iterable(this_labels_predicted)
```

Reply:
oh great! thanks for mentioning that @ritvikshrivastava